subscription.js

```diff
@@ -0,0 +1,181 @@
+    this.schema = schema
+    this.subscriber = subscriber
+    this.subscriptionContexts = new Map()
+
+    this.socket.on('message', this.handleMessage.bind(this))
+    this.socket.on('error', this.handleConnectionClose.bind(this))
+  }
+
+  async handleMessage (message) {
+    let data
+    try {
+      data = JSON.parse(message)
+    } catch (e) {
+      this.sendMessage(GQL_ERROR, null, 'Message must be a JSON string')
+      return
+    }
+
+    const { id, type } = data
+
+    switch (type) {
+      case GQL_CONNECTION_INIT:
+        this.sendMessage(GQL_CONNECTION_ACK)
+        break
+      case GQL_CONNECTION_TERMINATE:
+        this.handleConnectionClose()
+        break
+      case GQL_START:
+        try {
+          await this.handleGQLStart(data)
+        } catch (e) {
+          this.sendMessage(GQL_ERROR, id, e.message)
+        }
+        break
+      case GQL_STOP:
+        this.handleGQLStop(data)
+        break
+      default:
+        this.sendMessage(GQL_ERROR, id, 'Invalid payload type')
+    }
+  }
+
+  async handleGQLStart (data) {
+    // Starting a GraphQL subscription
+    const { id, payload } = data
+    const { query, variables, operationName } = payload
+
+    const sc = new SubscriptionContext({ subscriber: this.subscriber })
+    this.subscriptionContexts.set(id, sc)
+
+    const document = typeof query !== 'string' ? query : parse(query)
+
+    const result = await subscribe(
+      this.schema,
+      document,
+      {}, // rootValue
+      {
+        pubsub: sc
+      },
+      variables,
+      operationName
+    )
+
+    for await (const value of result) {
+      this.sendMessage(GQL_DATA, data.id, value)
+    }
+  }
+
+  handleGQLStop (data) {
+    const subscriptionContextToRemove = this.subscriptionContexts.get(data.id)
+    if (!subscriptionContextToRemove) {
+      return
+    }
+
+    subscriptionContextToRemove.close()
+    this.subscriptionContexts.delete(data.id)
+  }
+
+  handleConnectionClose () {
+    Array
+      .from(this.subscriptionContexts.values())
+      .map(subscriptionContext =>
+        subscriptionContext.close())
+    this.socket.close()
+  }
+
+  sendMessage (type, id, payload) {
+    this.socket.send(JSON.stringify({
+      type,
+      id,
+      payload
+    }))
+  }
+
+  close () {
+    this.handleConnectionClose()
+  }
+}
+
+function createConnectionHandler (schema, subscriber) {
+  return (connection, request) => {
+    connection.setEncoding('utf8')
+    const { socket } = connection
+
+    if (socket.protocol === undefined ||
+      (socket.protocol.indexOf(GRAPHQL_WS) === -1)) {
+      // Close the connection with an error code, ws v2 ensures that the
+      // connection is cleaned up even when the closing handshake fails.
+      // 1002: protocol error
+      socket.close(1002)
+
+      return
+    }
+
+    const subscriptionConnection = new SubscriptionConnection(socket, {
+      schema,
+      subscriber
+    })
+
+    connection.on('error', (e) => {})
```

Comment:
We should tear down the subscriptionConnection in case of an error, or are we doing it anyway?